src/plugins/workspace/public/workspace_client.ts

```diff
@@ -0,0 +1,275 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import {
+  HttpFetchError,
+  HttpFetchOptions,
+  HttpSetup,
+  WorkspaceAttribute,
+  WorkspacesSetup,
+} from '../../../core/public';
+
+const WORKSPACES_API_BASE_URL = '/api/workspaces';
+
+const join = (...uriComponents: Array<string | undefined>) =>
+  uriComponents
+    .filter((comp): comp is string => Boolean(comp))
+    .map(encodeURIComponent)
+    .join('/');
+
+type IResponse<T> =
+  | {
+      result: T;
+      success: true;
+    }
+  | {
+      success: false;
+      error?: string;
+    };
+
+interface WorkspaceFindOptions {
+  page?: number;
+  perPage?: number;
+  search?: string;
+  searchFields?: string[];
+  sortField?: string;
+  sortOrder?: string;
+}
+
+/**
+ * Workspaces is OpenSearchDashboards's visualize mechanism allowing admins to
+ * organize related features
+ *
+ * @public
+ */
+export class WorkspaceClient {
+  private http: HttpSetup;
+  private workspaces: WorkspacesSetup;
+
+  constructor(http: HttpSetup, workspaces: WorkspacesSetup) {
+    this.http = http;
+    this.workspaces = workspaces;
+  }
+
+  /**
+   * Initialize workspace list
+   */
+  public async init() {
+    await this.updateWorkspaceList();
+    this.workspaces.initialized$.next(true);
+  }
+
+  /**
+   * Add a non-throw-error fetch method for internal use.
+   */
+  private safeFetch = async <T = any>(
+    path: string,
+    options: HttpFetchOptions
+  ): Promise<IResponse<T>> => {
+    try {
+      return await this.http.fetch<IResponse<T>>(path, options);
+    } catch (error: unknown) {
+      if (error instanceof HttpFetchError) {
+        return {
+          success: false,
+          error: error.body?.message || error.body?.error || error.message,
+        };
+      }
+
+      if (error instanceof Error) {
+        return {
+          success: false,
+          error: error.message,
+        };
+      }
+
+      return {
+        success: false,
+        error: 'Unknown error',
+      };
+    }
+  };
+
+  private getPath(path: Array<string | undefined>): string {
+    return [WORKSPACES_API_BASE_URL, join(...path)].filter((item) => item).join('/');
+  }
+
+  private async updateWorkspaceList(): Promise<void> {
+    const result = await this.list({
+      perPage: 999,
+    });
+
+    if (result?.success) {
+      const resultWithWritePermission = await this.list({
```

Comment:
We don't have permission concepts in this PR, do we need to refactor here?

Reply:
You are right, refractored.